packages/util/src/__specs__/polymorphism.spec.tsx

```diff
@@ -0,0 +1,107 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+
+import { render } from '@testing-library/react'
+import React, { MouseEvent } from 'react'
+
+import { forwardRefWithAs, forwardRefWithAsAndStatics } from '..'
+
+interface Props {
+  optional?: boolean
+}
+const Button = forwardRefWithAs<Props, 'button'>((props, ref) => {
+  const { as: Comp = 'button', optional, ...rest } = props
+  return <Comp data-optional={optional} ref={ref} {...rest} />
+})
+
+const Anchor = forwardRefWithAs<Props, 'a'>((props, ref) => {
+  const { as: Comp = 'a', optional, ...rest } = props
+  return <Comp data-optional={optional} ref={ref} {...rest} />
+})
+
+interface Statics {
+  Anchor: typeof Anchor
+  Button: typeof Button
+}
+
+const CompoundComp = forwardRefWithAsAndStatics<Props, 'div', Statics>(
+  (props, ref) => {
+    const { as: Comp = 'div', optional, ...rest } = props
+
+    return <Comp data-optional={optional} ref={ref} {...rest} />
+  }
+)
+CompoundComp.Anchor = Anchor
+CompoundComp.Button = Button
+
+describe('polymorphism', () => {
+  describe('#forwardRefWithAs', () => {
+    it('renders without type errors', () => {
+      expect.assertions(0)
+
+      render(
+        <>
+          {/* Anchor accepts optional prop */}
+          <Anchor optional />
+
+          {/* Anchor accepts href prop */}
+          <Anchor href="#" />
+
+          {/* Button accepts optional prop */}
+          <Button optional />
+
+          {/* Button accepts disabled prop */}
+          <Button disabled />
+
+          {/* Button accepts onClick prop */}
+          <Button
+            onClick={(evt: MouseEvent<HTMLButtonElement>) =>
+              evt.currentTarget.disabled
+            }
+          />
+
+          {/* Button does NOT accept href prop */}
+          {/* @ts-expect-error */}
+          <Button href="#" />
+
+          {/* Button as "a" accepts href prop */}
+          <Button as="a" href="#" />
+
+          {/* Button as "a" accepts onClick prop */}
+          <Button
+            as="a"
+            onClick={(evt: MouseEvent<HTMLAnchorElement>) =>
+              evt.currentTarget.href
+            }
+          />
+
+          {/* Button as "Anchor" accepts href prop */}
+          <Button as={Anchor} href="#" />
```

Comment:
That's pretty cool that the types allow it. Not sure when I'd use though.

Reply:
maybe more of an advanced need. it's similar to our `render*` props where we allow passing in a ReactElement.

it also might help with composition using the various css-in-js libraries. here is the example i've tested with @emotion/styled - but i'm not sure if there is a better way to do this

```
export const Input = forwardRefWithAs<Props, "input">((props, ref) => {
  const { as: Comp = "input", children, ...rest } = props;

  return (
    <Comp ref={ref} {...rest}>
      {isFunction(children) ? children() : children}
    </Comp>
  );
});
```

```
import { Input } from 'some/ds/path'

const StyledInput = styled("input")`
  color: hotpink;
`;

export const CustomInput = (props) => <StyledInput as={Input} {...props} />;
```

there are some other ways to use emotion that i've tested here -> https://codesandbox.io/s/flamboyant-cloud-p7unf?file=/src/App.tsx